src/cinder/linux/GstPlayer.cpp

```diff
@@ -20,1298 +17,1147 @@
+
+    if( ! sEnableAsyncStateChange && stateChangeResult == GST_STATE_CHANGE_ASYNC ) {
+        g_print( " Blocking until pipeline state changes from : %s to %s\n", gst_element_state_get_name( current ), gst_element_state_get_name( targetState ) );
+        stateChangeResult = gst_element_get_state( mGstData.pipeline, &current, &pending, GST_CLOCK_TIME_NONE );
+    }
 	
-	return checkStateChange( stateChangeResult, targetState );
+    return checkStateChange( stateChangeResult );
 }
 
-bool GstPlayer::checkStateChange( GstStateChangeReturn stateChangeResult, const GstState & target )
+bool GstPlayer::checkStateChange( GstStateChangeReturn stateChangeResult )
 {
-	switch( stateChangeResult ) {
-		case GST_STATE_CHANGE_FAILURE:
-			g_warning( "**** Failed to change state.. ****" );
-			return false;
-			
-		case GST_STATE_CHANGE_SUCCESS:
-			g_print( "Pipeline state change SUCCESS from : %s to %s\n", gst_element_state_get_name( mGstData.mCurrentState ), gst_element_state_get_name ( target ) );
-			
-			mGstData.mTargetState = mGstData.mCurrentState = target;
-			
-			return true;;
-			
-		case GST_STATE_CHANGE_ASYNC:
-			mGstData.mTargetState = target;
-			mGstData.mCurrentState = getCurrentState();
-			mGstData.mAsyncStateChangePending = true;
-			g_print( "Pipeline state change ASYNC from : %s to %s\n", gst_element_state_get_name( mGstData.mCurrentState ), gst_element_state_get_name ( target ) );
-			
-			return true;;
-			
-		case GST_STATE_CHANGE_NO_PREROLL:
-			mGstData.mCurrentState = getCurrentState();
-			mGstData.mTargetState = target;
-			mGstData.mIsLive = true;
-			return true;
-			
-		default:
-			return false;
-	}
+    switch( stateChangeResult ) {
+        case GST_STATE_CHANGE_FAILURE: {
+	    g_warning( "** Pipeline failed to change state.. **" );
+            return false;	
+        }
+        case GST_STATE_CHANGE_SUCCESS: {
+            g_print( "Pipeline state changed SUCCESSFULLY from : %s to %s\n", gst_element_state_get_name( mGstData.currentState ), gst_element_state_get_name ( mGstData.targetState ) );
+            // Our target state is now the current one also.
+            mGstData.updateState( mGstData.targetState );
+            return true;
+        }	
+        case GST_STATE_CHANGE_ASYNC: {
+            g_print( "Pipeline state change will happen ASYNC from : %s to %s\n", gst_element_state_get_name( mGstData.currentState ), gst_element_state_get_name ( mGstData.targetState ) );
+            return true;
+        }
+        case GST_STATE_CHANGE_NO_PREROLL: {
+            mGstData.isLive = true;
+            return true;
+        }
+        default:
+            return false;
+    }
 }
 
-GstData& GstPlayer::getGstData()
+void GstPlayer::createTextureFromMemory()
 {
-	return mGstData;
+    mMutex.lock();
+    std::swap( mFrontVBuffer, mBackVBuffer );
+    mMutex.unlock();

```

Comment:
Same threading question here, if the unlock needs to happen after `mFrontBuffer` is used below.